src/sentry/preprod/api/endpoints/organization_preprod_artifact_assemble.py

```diff
@@ -130,6 +130,24 @@ def post(self, request: Request, project: Project) -> Response:
             checksum = data.get("checksum")
             chunks = data.get("chunks", [])
 
+            # Validate VCS parameters - if any are provided, all required ones must be present
+            vcs_params = {
+                "head_sha": data.get("head_sha"),
+                "head_repo_name": data.get("head_repo_name"),
+                "provider": data.get("provider"),
+                "head_ref": data.get("head_ref"),
+            }
+            provided_params = {k: v for k, v in vcs_params.items() if v}
+
+            if provided_params and len(provided_params) != len(vcs_params):
```

Comment:
Here I think you don't need `provided_params and` but also I might be cleaner to write this as:

```python

if any(not v for v in vcs_params.values()):
  # ...
```